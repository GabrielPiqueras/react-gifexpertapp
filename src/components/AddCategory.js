import React, { useState } from 'react';
import PropTypes from 'prop-types';

/**
 * COMPONENTE ADDCATEGORY
 * 
 * 1. Muestra el estado del componente (inputValue) en la caja de texto.
 * 2. Si cambia el valor de la caja de texto, dicho valor será el nuevo estado.
 * 3. Si se envía el formulario y el texto tiene al menos 2 letras, se añade la categoría al array.
 * 4. Al recibir la función como propiedad que cambia el valor del estado del componente GifExpertApp (setCategories), 
 *    podemos añadir el nuevo valor al array.
 * 5. Se reinicia el input para dejarlo vacío.
 * 
 */
const AddCategory = ({ setCategories }) => {
        
    const [ inputValue, setInputValue ] = useState('');

    const handleInputChange = (e) => {
        setInputValue(e.target.value);
    }

    const handleSubmit = (e) => {
        e.preventDefault();

        if (inputValue.trim().length > 2) {
            setCategories( cats => [inputValue, ...cats ]);
            setInputValue('');
        }
    }

    return(
        <form onSubmit={ handleSubmit }>
            <input
                onChange={ handleInputChange }
                id='category'
                type='text'
                value={ inputValue }
                placeholder='Search'
            />
        </form>
    )
}

AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired,
}

export default AddCategory;