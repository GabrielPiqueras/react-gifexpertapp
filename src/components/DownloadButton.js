import React from 'react';
import { ArrowDownTrayIcon } from '@heroicons/react/20/solid'

const DownloadButton = ({ gifUrl, fileName }) => {
    
  const handleDownload = async () => {
    const response = await fetch(gifUrl);
    const blob = await response.blob();
    const url = window.URL.createObjectURL(blob);

    const link = document.createElement('a');
    link.href = url;
    link.download = fileName;
    document.body.appendChild(link);
    link.click();
    document.body.removeChild(link);
    window.URL.revokeObjectURL(url);
  };

  return (
    <button
      className="flex justify-end items-start ml-auto bg-gray-500 hover:bg-gray-400 text-white font-bold py-2 px-2 border-b-4 border-gray-700 hover:border-gray-500 rounded"
      onClick={handleDownload}
      title='Descargar Gif'
    >
      <ArrowDownTrayIcon className="w-4 h-4"/>
    </button>
  );
}

export default DownloadButton;