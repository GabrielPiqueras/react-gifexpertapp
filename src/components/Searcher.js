import React, { useState } from 'react';
import PropTypes from 'prop-types';

import { MagnifyingGlassIcon } from '@heroicons/react/20/solid'

/**
 * COMPONENTE ADDCATEGORY
 * 
 * 1. Muestra el estado del componente (inputValue) en la caja de texto.
 * 2. Si cambia el valor de la caja de texto, dicho valor será el nuevo estado.
 * 3. Si se envía el formulario y el texto tiene al menos 2 letras, se añade la categoría al array.
 * 4. Al recibir la función como propiedad que cambia el valor del estado del componente GifSearcher (setPrompt), 
 *    podemos añadir el nuevo valor al array.
 * 5. Se reinicia el input para dejarlo vacío.
 * 
 */
export const Searcher = ({ setPrompt }) => {
        
    const [ inputValue, setInputValue ] = useState('');

    const handleInputChange = (e) => {
        setInputValue(e.target.value);
    }

    const handleSubmit = (e) => {
        e.preventDefault();

        if (inputValue.trim().length > 2) {
            setPrompt(inputValue);
            setInputValue('');
        }
    }

    return(
        <form onSubmit={ handleSubmit }>

            <div className="mt-12 mb-4 flex justify-center">
                <input
                    autoFocus 
                    id='category'
                    type='text'
                    value={ inputValue }
                    onChange={ handleInputChange }
                    style={{width: '35%'}}
                    className="shadow appearance-none border rounded w-full py-2 px-3 text-gray-700 leading-tight focus:outline-none focus:shadow-outline w-64 h-12 mr-4"
                    placeholder="Gifs de..."
                />
                <button
                    className="bg-blue-500 hover:bg-blue-400 text-white font-bold py-2 px-4 border-b-4 border-blue-700 hover:border-blue-500 rounded flex items-center"
                    onClick={ handleSubmit }
                >
                    <MagnifyingGlassIcon className="w-5 h-5 mr-1"/> Buscar
                </button>
            </div>
        </form>
    )
}

Searcher.propTypes = {
    setPrompt: PropTypes.func.isRequired,
}