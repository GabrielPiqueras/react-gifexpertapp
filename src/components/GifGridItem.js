import React from 'react'
import PropTypes from 'prop-types';
import DownloadButton from './DownloadButton';

export const GifGridItem = ({ id, title, url }) => {

  const convertToSlug =(title)  => {
    const slug = title
      .toLowerCase()
      .replace(/[^\w\s-]/g, '') // Elimina caracteres especiales
      .replace(/\s+/g, '-') // Reemplaza espacios en blanco con guiones
      .replace(/--+/g, '-'); // Elimina guiones consecutivos
    
    return slug;
  }

  return (
    <>
      {/* Nueva */}
      <div className="bg-white max-w-sm rounded overflow-hidden shadow-lg card animate__animated animate__fadeIn m-4 w-1/1 sm:w-1/2 md:w-1/3 lg:w-1/4 xl:w-1/3">
        <img className="w-full" src={ url } alt={ title} />
        <div className="py-4">
          <div className="pl-6 pr-4 font-bold text-xl mb-2 flex justify-end items-start">
            { title }
            <DownloadButton gifUrl={ url } fileName={ convertToSlug(title) } />
            
            </div>
          
        </div>
        <div>
          <p className="px-6 text-gray-700 text-base">
            Lorem ipsum dolor sit amet, consectetur adipisicing elit. Voluptatibus
            quia, nulla! Maiores et perferendis eaque, exercitationem praesentium
            nihil.
          </p>
        </div>
        <div className="px-6 pt-4 pb-2">
          <span className="inline-block bg-gray-200 rounded-full px-3 py-1 text-sm font-semibold text-gray-700 mr-2 mb-2">
            #gif
          </span>
          <span className="inline-block bg-gray-200 rounded-full px-3 py-1 text-sm font-semibold text-gray-700 mr-2 mb-2">
            #hastag
          </span>
          <span className="inline-block bg-gray-200 rounded-full px-3 py-1 text-sm font-semibold text-gray-700 mr-2 mb-2">
            #other
          </span>

            

          {/* <button class="bg-blue-500 hover:bg-blue-400 text-white font-bold py-2 px-4 border-b-4 border-blue-700 hover:border-blue-500 rounded">
            Descargar GIF
          </button> */}
         
        </div>
      </div>
    </>
  )
}

GifGridItem.propTypes = {
  title: PropTypes.string.isRequired,
  url: PropTypes.string.isRequired
}