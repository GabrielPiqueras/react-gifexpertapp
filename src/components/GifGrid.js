import React from 'react';
import { useFetchGifs } from '../hooks/useFetchGifs';
import { GifGridItem } from './GifGridItem';

import PropTypes from 'prop-types';

// import { getGifs } from '../helpers/getGifs';

/**
 * COMPONENTE GIFGRID
 * 
 * - Muestra los gifs según el término buscado en giphy, son mostrados en un div con sus títulos.
 * 
 * 1. El estado del componente será un array (vacío por defecto) donde se guardarán los gifs
 * 2. Los gifs se obtienen usando la Api de Giphy
 * 3. Se buscan los gifs poniendo mi ApiKey de la plataforma, el límite de gifs y el término a buscar, que será la categoría
 *    recibida por parámetro. Dichas categorías se guardan en un array del componente principal.
 * 4. Al recibir los gifs en la variable data, la mapeo para extraer el id, titulo e imagen de cada gif.
 * 5. Una vez hecho el mapeo, se pone dicho array como valor del estado.
 * 6. Usamos el Hook useEffect() para que solo haga la petición una vez al principio, y no si se recarga el componente.
 * 7. Pinto los gifs del array en html, con sus titulos e imagenes.
 */

export const GifGrid = ({ category }) => {

  // Uso mi hook personalizado que trae los gifs
  const { data: gifs, loading } = useFetchGifs(category);

  return (
    <>
      <h3>{ category }</h3>

      {/* Más práctico de mostrar */}
      { loading && <p>Cargando...</p> } 
      {/* { loading ? <p>Cargando...</p>: <p>Otro mensaje.</p> } */}

      <div className='card-grid'>
          {
            // Usando el {...} sería como enviar el id, title y url como propiedades por separado (otra forma de hacerlo) en vez de pasarlo dentro de img
            gifs.map(gif =>
              <GifGridItem key={ gif.id } {...gif} />
              // <GifGridItem key={ gif.id } img={ gif } />
            )
          }
      </div>
    </>
  )
}

GifGrid.propTypes = {
  category: PropTypes.string.isRequired, 
} 
