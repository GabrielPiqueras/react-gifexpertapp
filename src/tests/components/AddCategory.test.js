import React from 'react';
import { shallow } from 'enzyme';

import '@testing-library/jest-dom';
import { AddCategory } from '../../components/AddCategory';

describe('Pruebas en <AddCategory />', () => {

    const setCategories = jest.fn();
    
    let wrapper = shallow( <AddCategory setCategories={ setCategories } /> );

    beforeEach( () => {
        jest.clearAllMocks();
        wrapper = shallow( <AddCategory setCategories={ setCategories } /> );
    });

    test('<AddCategory /> debe de mostrarse correctamente', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('debe de cambiar la caja de texto', () => {
        const input = wrapper.find('input');
        const value = 'Hola mundo';

        input.simulate('change', {target: { value } });
        const inputValue = wrapper.find('input').prop('value');

        expect(inputValue).toBe(value);
    });

    test('NO debe de postear la información con submit', () => {
        wrapper.find('form').simulate('submit', { preventDefault(){} }); // Detenemos el evento con preventDefalt()
        expect( setCategories ).not.toHaveBeenCalled(); // No debe haber sido llamado
    });

    test('debe de llamar el setCategories y limpiar la caja de texto', () => {

        // Simulo change
        const value = 'Bobobo';
        wrapper.find('input').simulate('change', {target: { value }})

        // Simulo el submit del formulario
        wrapper.find('form').simulate('submit', { preventDefault(){} });
        
        // SetCategories debe haber sido llamado
        expect( setCategories ).toHaveBeenCalled(); // que haya sido llamada
        expect( setCategories ).toHaveBeenCalledTimes(1); // que solo se llamó una vez
        expect( setCategories ).toHaveBeenCalledWith( expect.any(Function)  ); // que haya sido llamada con una función (any = cualquier función)
        
        // El valor del input debe estar vacío
        const inputValue  = wrapper.find('input').prop('value');

        expect(inputValue).toBe('');
    })
})