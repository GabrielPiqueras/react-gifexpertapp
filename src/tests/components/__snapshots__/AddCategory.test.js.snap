// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <AddCategory /> <AddCategory /> debe de mostrarse correctamente 1`] = `
<form
  onSubmit={[Function]}
>
  <input
    id="category"
    onChange={[Function]}
    placeholder="Search"
    type="text"
    value=""
  />
</form>
`;
